examples/tls-rustls/src/main.rs

```diff
@@ -42,3 +54,25 @@ async fn main() {
 async fn handler() -> &'static str {
     "Hello, World!"
 }
+
+async fn redirect() {
+    let addr = SocketAddr::from(([127, 0, 0, 1], 7878));
+    tracing::debug!("http redirect listening on {}", addr);
+
+    axum::Server::bind(&addr)
+        .serve(fallback.into_make_service())
+        .await
+        .unwrap()
+}
+
+async fn fallback(Host(host): Host, uri: Uri) -> Redirect {
+    tracing::debug!("308: Permanent Redirect");
+
+    // Can remove call to `.replace()` if using default
+    // ports for http (80) and https (443).
+    Redirect::permanent(&*format!(
+        "https://{}{}",
+        &host.replace("7878", "3000"),
+        &uri
+    ))
+}
```

Comment:
What do you think about doing something like this instead:

```suggestion
#[derive(Clone, Copy)]
struct Ports {
    http: u16,
    https: u16,
}

async fn redirect_http_to_https(ports: Ports) {
    fn make_https(host: String, uri: Uri, ports: Ports) -> Result<Uri, BoxError> {
        let mut parts = uri.into_parts();

        parts.scheme = Some(axum::http::uri::Scheme::HTTPS);

        if parts.path_and_query.is_none() {
            parts.path_and_query = Some("/".parse().unwrap());
        }

        let https_host = host.replace(&ports.http.to_string(), &ports.https.to_string());
        parts.authority = Some(https_host.parse()?);

        Ok(Uri::from_parts(parts)?)
    }

    let redirect = move |Host(host): Host, uri: Uri| async move {
        match make_https(host, uri, ports) {
            Ok(uri) => Ok(Redirect::permanent(&uri.to_string())),
            Err(error) => {
                tracing::warn!(%error, "failed to convert URI to HTTPS");
                Err(StatusCode::BAD_REQUEST)
            }
        }
    };

    let addr = SocketAddr::from(([127, 0, 0, 1], ports.http));
    tracing::debug!("http redirect listening on {}", addr);

    axum::Server::bind(&addr)
        .serve(redirect.into_make_service())
        .await
        .unwrap()
}
```

So
- Make a `Ports` struct instead of duplicating the ports
- Using `Uri::into_parts` and handling errors, rather than working with strings

Reply:
Sorry I didn't see the "Commit suggestion" button.  Next time I'll know to use it.